lib/backpex/fields/has_many.ex

```diff
@@ -381,13 +332,155 @@ defmodule Backpex.Fields.HasMany do
+    |> maybe_offset_query(Keyword.get(opts, :offset))
+    |> maybe_limit_query(Keyword.get(opts, :limit))
+    |> select([x], {field(x, ^display_field_form(field)), x.id})
+    |> repo.all()
+  end
+
+  defp maybe_limit_query(query, nil), do: query
+  defp maybe_limit_query(query, limit), do: query |> limit(^limit)
+
+  defp maybe_offset_query(query, nil), do: query
+  defp maybe_offset_query(query, offset), do: query |> offset(^offset)
+
+  defp maybe_options_query(query, %{options_query: options_query}, assigns), do: options_query.(query, assigns)
+  defp maybe_options_query(query, _field_options, _assigns), do: query
+
+  defp maybe_search_query(query, _schema_name, _field_options, _display_field, nil), do: query
+
+  defp maybe_search_query(query, schema_name, field_options, display_field, search_input) do
+    if String.trim(search_input) == "" do
+      query
+    else
+      search_input = "%#{search_input}%"
+      select = Map.get(field_options, :select, nil)
+
+      if select do
+        where(query, ^dynamic(ilike(^select, ^search_input)))
+      else
+        where(query, [{^schema_name, schema_name}], ilike(field(schema_name, ^display_field), ^search_input))
+      end
+    end
+  end
+
+  defp count_options(assigns, opts \\ []) do
+    %{schema: schema, repo: repo, field: field, field_options: field_options, name: name} = assigns
+
+    display_field = display_field(field)
+
+    %{queryable: queryable} = schema.__schema__(:association, name)
+    schema_name = Resource.name_by_schema(queryable)
+
+    from(queryable, as: ^schema_name)
+    |> maybe_options_query(field_options, assigns)
+    |> maybe_search_query(schema_name, field_options, display_field, Keyword.get(opts, :search))
+    |> subquery()
+    |> repo.aggregate(:count, :id)
+  end
+
+  defp assign_selected(socket) do
+    %{
+      assigns:
+        %{
+          form: form,
+          field: field,
+          name: name,
+          schema: schema,
+          repo: repo,
+          options_count: options_count,
+          field_options: field_options,
+          options: options
+        } = assigns
+    } = socket
+
+    %{queryable: queryable} = schema.__schema__(:association, name)
+
+    selected_ids = extract_selected_ids(form[name].value)
+
+    options_map = Map.new(options, fn {label, id} -> {id, {label, id}} end)
+
+    {selected_from_options, ids_to_fetch} =
+      Enum.reduce(selected_ids, {[], []}, fn id, {from_options, to_fetch} ->
+        case Map.get(options_map, id) do
+          nil -> {from_options, [id | to_fetch]}
+          item -> {[item | from_options], to_fetch}
+        end
+      end)
+
+    selected_from_db =
+      if Enum.empty?(ids_to_fetch) do
+        []
+      else
+        queryable
+        |> where([x], x.id in ^ids_to_fetch)
+        |> maybe_options_query(field_options, assigns)
+        |> select([x], {field(x, ^display_field_form(field)), x.id})
+        |> repo.all()
+      end
```

Comment:
Should be its own functions to improve readability here.

You also don't need the big pattern matching block at the start of the function when splitting things up. It is also OK to write `assigns.foo` in some places, I think.